sbi/inference/base.py

```diff
@@ -0,0 +1,128 @@
+
+from abc import ABC
+from typing import Optional, Callable
+import warnings
+import os.path
+import torch
+from torch.utils.tensorboard import SummaryWriter
+from torch import Tensor
+from torch.distributions import Uniform
+from sbi.utils.torchutils import atleast_2d, get_default_device
+from sbi.simulators.simutils import set_simulator_attributes
+from sbi.utils import get_log_root, get_timestamp
+from copy import deepcopy
+from sbi.inference.posteriors.sbi_posterior import Posterior
+
+
+class NeuralInference(ABC):
+    """Abstract base class for neural inference methods.
+    """
+
+    def __init__(
+        self,
+        simulator: Callable,
+        prior: torch.distributions.Distribution,
+        true_observation: Tensor,
+        simulation_batch_size: int = 1,
+        device: Optional[torch.device] = None,
+        summary_writer: Optional[SummaryWriter] = None
+    ):
+
+        """
+        Args:
+            simulator: function that takes (vector) parameter and returns one  
+                (vector) result per parameter vector, with shape (num_parameters, result_dimension)
+            
+            prior: distribution-like object with `log_prob`and `sample` methods.
+            
+            true_observation: tensor containing the observation x0.
+            
+            simulation_batch_size: number of parameter vectors that the 
+                 simulator accepts and converts to data x at once. If -1, we simulate all parameter sets at the same time. If >= 1, the simulator has to process data of shape (sim_batch_size, parameter_dimension).
+            
+            summary_writer: an optional SummaryWriter to control log file 
+                location (default is <current working directory>/logs.)
+             
+            device: torch.device on which to compute (optional).
+            
+            mcmc_method: MCMC method to use for posterior sampling, one of 
+                ['slice', 'hmc', 'nuts'].
+        """
+
+        # check inputs: observation
+        self._true_observation = atleast_2d(true_observation)
+
+        # warn if prior is Uniform and dim_input > 1 (inferred by drawing once)
+        # XXX if prior is a torch Distribution, then it should support
+        # XXX event_shape, no need to sample
+        # XXX also, why use 'input' here? -> parms are input to the simulator
+        scalar_input = prior.sample().numel() == 1
```

Comment:
One reason to sample might be that it's slightly more general and cheap to do anyway. Not that we are doing this in any example application for now, but in theory the prior might be a probabilistic program

Reply:
Yes but in this context the point is totally moot because the warning only triggers if the distribution is a `torch.distributions.Distribution.Uniform`, so I put this in a shortcircuiting `and` clause and used `batch_shape`, which is even clearer (upcoming round of the PR)